lib/mongo/server/connection_pool.rb

```diff
@@ -73,19 +119,79 @@ def checkin(connection)
       #
       # @since 2.0.0
       def checkout
-        queue.dequeue
+        raise_if_closed!
+
+        publish_cmap_event(
+          Monitoring::Event::ConnectionCheckoutStarted.new(address)
+        )
+
+        deadline = Time.now + wait_timeout
+        @wait_queue.enter_wait_queue(wait_timeout, deadline) do
+          connections.pop(deadline).tap do |c|
+            publish_cmap_event(
```

Comment:
It seems to me that this event publication should not be done while the thread is considered to be in the wait queue.

Reply:
Fair enough; I'll move that to outside the block